streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalSegmentIterator.java

```diff
@@ -0,0 +1,138 @@
+        this.snapshotOwner = null;
+    }
+
+    @Override
+    public void close() {
+        // user may refuse consuming all returned records, so release the snapshot when closing the iterator if it is not released yet!
+        releaseSnapshot();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return iterator.hasNext() || maybeFillIterator();
+    }
+
+    @Override
+    public Object next() {
+        if (hasNext()) {
+            return iterator.next();
+        }
+        return null;
+    }
+    private boolean maybeFillIterator() {
+
+        final List<VersionedRecord<byte[]>> queryResults = new ArrayList<>();
+        while (segmentIterator.hasNext()) {
+            final LogicalKeyValueSegment segment = segmentIterator.next();
+
+            if (snapshot == null) { // create the snapshot (this will happen only one time).
+                this.snapshotOwner = segment;
+                // take a RocksDB snapshot to return the segments content at the query time (in order to guarantee consistency)
+                final Lock lock = new ReentrantLock();
+                lock.lock();
+                try {
+                    this.snapshot = snapshotOwner.getSnapshot();
+                } finally {
+                    lock.unlock();
+                }
+            }
+
+            final byte[] rawSegmentValue = segment.get(key, snapshot);
+            if (rawSegmentValue != null) { // this segment contains record(s) with the specified key
+                if (segment.id() == -1) { // this is the latestValueStore
+                    final long recordTimestamp = RocksDBVersionedStore.LatestValueFormatter.getTimestamp(rawSegmentValue);
+                    if (recordTimestamp <= toTime) {
+                        // latest value satisfies timestamp bound
+                        queryResults.add(new VersionedRecord<>(RocksDBVersionedStore.LatestValueFormatter.getValue(rawSegmentValue), recordTimestamp));
+                    }
+                } else {
+                    // this segment contains records with the specified key and time range
+                    final List<RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult> searchResults =
+                            RocksDBVersionedStoreSegmentValueFormatter.deserialize(rawSegmentValue).findAll(fromTime, toTime);
+                    for (final RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult searchResult : searchResults) {
+                        queryResults.add(new VersionedRecord<>(searchResult.value(), searchResult.validFrom(), Optional.of(searchResult.validTo())));
+                    }
+                }
+            }
+            if (!queryResults.isEmpty()) {
+                break;
+            }
+        }
+        if (!queryResults.isEmpty()) {
+            if (order.equals(ResultOrder.ASCENDING)) {
+                queryResults.sort((r1, r2) -> (int) (r1.timestamp() - r2.timestamp()));
```

Comment:
Wondering if we need to sort? Could we also travers `queryResult` backwards later instead?

It seems we could init with `queryResults.listIterator(queryResult.size())` and use `iterator.previous()` instead of `iterator.next()` inside `next()` ?